src/canary.rs

```diff
@@ -186,3 +185,88 @@ impl Canary {
         }
     }
 }
+
+/// Write [`CANARY_VALUE`] to the stack.
+fn paint_stack(core: &mut Core, start: u32, mut end: u32) -> Result<(), probe_rs::Error> {
```

Comment:
I'd add a note here to observe that `start` should be the numerically lower address, and `end` should be the numerically higher address: that is `start < end`.

Reply:
Maybe also panic if that is not upheld. Otherwise the subtraction at line 197 will underflow and panic.

Reply:
Does `start` also have to be aligned?

Reply:
> I'd add a note here to observe that `start` should be the numerically lower address, and `end` should be the numerically higher address: that is `start < end`.

Good point, I am asserting this in line 250 in `fn subroutine`, but it is probably better to already check it before.



> Does `start` also have to be aligned?

I think so, but I was not sure how to go about it if it is not. Probably subtracting the modulo, right?

Reply:
> Good point, I am asserting this in line 250 in fn subroutine, but it is probably better to already check it before.

I'd apply the comment to whichever functions take `start` and `end` as arguments :)

Reply:
> Probably subtracting the modulo, right?

Yes, I think so. Or just bailing out - the linker script really should have set the stack start and end to be 4-byte aligned anyway, plus https://developer.arm.com/documentation/ddi0403/d/System-Level-Architecture/System-Level-Programmers--Model/ARMv7-M-exception-model/Stack-alignment-on-exception-entry says:


> The ARMv7-M architecture guarantees that stack pointer values are at least 4-byte aligned.

Reply:
Interesting. With the nrf52840-dk and the `nrf52840-hal` I observe the `end` to be not 4-, but only 2-byte-aligned.

Before aligning it:
```text
start: 0x20000000, start % 4 == 0
end: 0x2003FBC6, end % 4 == 2
```
After aligning it: 
```
start: 0x20000000, start % 4 == 0
end: 0x2003FBC8, end % 4 == 0
```

Reply:
The value of `end` should be the initial stack pointer minus one.

Reply:
My only concern is, since the initial stack pointer points to the address directly _above_ the stack, it itself doesn't belong to the stack anymore. Can anything bad happen when we are writing to this address? (I don't think so because static variables and so on will only be created later, but I still want to double check).

Reply:
If the stack is at the top of the RAM, the initial stack pointer value will be an invalid address.

From `memory.x` in `cortex-m-rt`:

> `_stack_start = ORIGIN(CCRAM) + LENGTH(CCRAM);`

The problem, I think, is a range doing `-1` when we should be doing `-4`, because we work in 4-byte units. Or perhaps we shouldn't hold *start* and *end* and instead hold *start* and *length-in-32-bit-words*.

Reply:
> The problem, I think, is a range doing `-1` when we should be doing `-4`, because we work in 4-byte units. Or perhaps we shouldn't hold *start* and *end* and instead hold *start* and *length-in-32-bit-words*.

So changing it to `initial_stack_pointer - 4` should be fine? 
I'd like to avoid bigger refactorings of other parts of the code as part of this PR.

Reply:
See https://github.com/knurling-rs/probe-run/pull/302/commits/67c15760a1eeb13f006ec371562917c36c16dfa5.